### PR TITLE
Update user logging sequence and make collections deal with reactivity better

### PR DESCRIPTION
**Pull request by ToyboxZach**
Restructure how we handle the Tracker, instead of a global listen to any change make the listens happen on specific queries and collections. Also fixes some bugs where we would generate way too many instances of subscriptions.

This also fixes the log in flow so the series of loggingIn -> LoggedIN is consistent and doesn't allow the package to ever auto log you out as that should be handled by the owning app just like in base meteor.

Every time you register a callback it gets added to a generic "changed" callback, that means that every single callback needs to be called on every single change to a collection which causes a huge hold up on the UI thread.

To make it even worst the subscriptions don't get cleaned up when you resubscribe so every time a "withTracker" or "useTracker" was called you would get stacked subscriptions. This then would lead to 100 plus subscriptions for a single component after some time which leads to bigger and bigger slow downs.

My PR separates that out as much as possible and makes sure that you only have as many subscriptions as you actually need. It is a pretty hefty change and I don't personally use and local collections in my app so those are going to be pretty poorly tested, but if other people want to take my change as a jumping off point I think its a pretty good point for server only subscriptions.

This also fixes up weirdness with the logging in and order of events where you would have an invalid user state or it would force you to logout because of bad internet

Likely fixes these erorrs:
https://github.com/meteorrn/meteor-react-native/issues/75
https://github.com/meteorrn/meteor-react-native/issues/58

and maybe
https://github.com/meteorrn/meteor-react-native/issues/79

### CODE REVIEW

File: src/Data.js [hunk 1]
```diff
@@ -71,11 +71,13 @@ export default {
     );
   },
   notify(eventName) {
-    this._cbs.map((cb) => {
-      if (cb.eventName == eventName && typeof cb.callback == 'function') {
-        cb.callback();
-      }
-    });
+    setTimeout(() => {
```
Comment (jankapunkt) [inline src/Data.js:74]:
```js
// what this block does
// why it's needed
```

File: src/Meteor.js [hunk 2]
```diff
@@ -111,6 +120,14 @@ const Meteor = {
           }
         }
       );
+
+      NetInfo.fetch().then(
```
Comment (jankapunkt) [inline src/Meteor.js:124]:
```js
// what this block does
// why it's needed
```

File: src/components/useTracker.js [hunk 3]
```diff
@@ -1,33 +1,44 @@
-import { useEffect, useState } from 'react';
+import { useEffect, useRef, useReducer, useMemo } from 'react';
 import Tracker from '../Tracker.js';
-import Data from '../Data';
 
+const fur = (x: number): number => x + 1;
+const useForceUpdate = () => useReducer(fur, 0)[1];
 export default (trackerFn, deps = []) => {
```
Comment (jankapunkt) [inline src/components/useTracker.js:6]: Please add some documentation for this module, especially how it behaves, preferrably jsDoc; optimally would be to add comments on each of the React hooks so other devs understand how they work and why they were chosen

Comment (ToyboxZach) [reply src/components/useTracker.js:6]: This file was copied from meteor-react-data at some point, I don't really have an amazing understanding enough to do anything with this.

**Review by jankapunkt — APPROVED** [review]
Took me a while since we released yesterday our app but I worked through the code and on my end this did not introduce any issues. From what I see in the code this has potential to be a good starting point for even further optimization.

I will prepare 2.3.0 for publish.

**Pull request merged**